bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryParam.scala

```diff
@@ -0,0 +1,233 @@
+      case "true" => YesNoType.YES.displayValue()
+      case "false" => YesNoType.NO.displayValue()
+      case _ => expression
+    }
+
+    override def matches(expression: String, x: A): Boolean = (for {
+      r <- Option(expression).map(transform).map(_.toRegex)
+      m <- Option(x).flatMap(extractor).map(_.displayValue)
+    } yield r.findFirstMatchIn(m).isDefined).getOrElse(false)
+  }
+}
+
+
+sealed case class LchQueryParam[A](name: String, v: ValueMatcher[A])
+
+object LchQueryParam {
+  import ValueMatcher._
+
+  private[servlet] implicit class ToYesNo(val b: Boolean) extends AnyVal {
+    def toYesNo: YesNoType =
+      b ? YesNoType.YES | YesNoType.NO
+  }
+
+
+  private[servlet] implicit class ISPProgramExtractors(val prog: ISPProgram) extends AnyVal {
+    def toSPProg: Option[SPProgram] =
+      Option(prog).map(_.getDataObject.asInstanceOf[SPProgram])
+
+    def semester: Option[String] = for {
+      s <- Option(prog.getProgramID).map(_.stringValue)
+      i = s.indexOf('-')
+      if i != -1 && s.length > i+6 && s.charAt(i+1) == '2'
+    } yield s.substring(i+1, i+6)
+  }
+
+
+  private[servlet] implicit class ISPObservationExtractors(val obs: ISPObservation) extends AnyVal {
+    def toSPObs: Option[SPObservation] =
+      Option(obs).map(_.getDataObject.asInstanceOf[SPObservation])
+
+    def instrumentType: Option[String] =
+      SPTreeUtil.findInstruments(obs).asScala.map(_.getDataObject).collect {
+        case inst: SPInstObsComp => inst.getType.readableStr
+      }.headOption
+
+    def ao: AOConstants.AO = {
+      Option(AOTreeUtil.findAOSystem(obs)).map(_.getDataObject).
+        collect {
+          case inst: InstAltair if inst.getGuideStarType == AltairParams.GuideStarType.LGS => AOConstants.AO.Altair_LGS
+          case inst: InstAltair if inst.getGuideStarType == AltairParams.GuideStarType.NGS => AOConstants.AO.Altair_NGS
+        }.getOrElse(AOConstants.AO.NONE)
+    }
+
+    def targetEnvironment: Option[TargetEnvironment] = for {
+      n <- Option(SPTreeUtil.findTargetEnvNode(obs))
+      c <- Option(n.getDataObject).map(_.asInstanceOf[TargetObsComp])
+      t <- Option(c.getTargetEnvironment)
+    } yield t
+
+    def timingWindows: Option[List[SPSiteQuality.TimingWindow]] = for {
+      n <- Option(SPTreeUtil.findObsCondNode(obs))
+      c <- Option(n.getDataObject).map(_.asInstanceOf[SPSiteQuality])
+    } yield c.getTimingWindows.asScala.toList
+  }
+
+
+
+  private[servlet] val ProgramSemesterParam = LchQueryParam("programSemester",
+    new StringValueMatcher[ISPProgram] {
+      override protected def extractor(prog: ISPProgram): Option[String] =
+        prog.semester
+    }
+  )
+
+  private[servlet] val ProgramTitleParam = LchQueryParam("programTitle",
+    new StringValueMatcher[ISPProgram] {
+      override protected def extractor(prog: ISPProgram): Option[String] =
+        prog.toSPProg.map(_.getTitle)
+    }
+  )
```

Comment:
Now parameters, instead of being mere `String`s that are manually extracted, are self-contained `LchQueryParams` with their `name`, and the `ValueMatcher` that extracts the value from the appropriate data object for comparison with the supplied pseudo-RE as per the ODB Browser page.